File: wallet-kit/src/main/java/bitcoin/walllet/kit/network/message/PingMessage.java
package bitcoin.walllet.kit.network.message;

import java.io.ByteArrayInputStream;
import java.io.IOException;

import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.common.io.BitcoinOutput;
import bitcoin.walllet.kit.common.util.RandomUtils;

/**
 * Build P2P message:
 * https://en.bitcoin.it/wiki/Protocol_documentation#Message_structure
 *
 * @author liaoxuefeng
 */
public class PingMessage extends Message {

    long nonce;

    public PingMessage() {
        super("ping");
        this.nonce = RandomUtils.randomLong();
    }

    public PingMessage(byte[] payload) throws IOException {
        super("ping");
        try (BitcoinInput input = new BitcoinInput(new ByteArrayInputStream(payload))) {
            this.nonce = input.readLong();
        }
    }

    public long getNonce() {
        return this.nonce;
    }

    @Override
    protected byte[] getPayload() {
        return new BitcoinOutput().writeLong(this.nonce).toByteArray();
    }

    @Override
    public String toString() {
        return "PingMessage(nonce=" + this.nonce + ")";
    }

}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/messages/PingMessage.kt
package bitcoin.wallet.kit.messages

import bitcoin.walllet.kit.io.BitcoinInput
import bitcoin.walllet.kit.io.BitcoinOutput
import bitcoin.walllet.kit.utils.RandomUtils
import java.io.ByteArrayInputStream
import java.io.IOException

/**
 * Ping Message
 *
 *  Size        Field   Description
 *  ====        =====   ===========
 *  8 bytes     Nonce   Random value
 */
class PingMessage : Message {

    var nonce: Long = 0
        internal set

    constructor() : super("ping") {
        nonce = RandomUtils.randomLong()
    }

    @Throws(IOException::class)
    constructor(payload: ByteArray) : super("ping") {
        BitcoinInput(ByteArrayInputStream(payload)).use { input ->
            nonce = input.readLong()
        }
    }

    override fun getPayload(): ByteArray {
        return BitcoinOutput().writeLong(nonce).toByteArray()
    }

    override fun toString(): String {
        return "PingMessage(nonce=$nonce)"
    }
}
